### DISCUSSION TITLE
tiffsave fails with CreateFileMapping failed and/or TiffAppendToStrip exceptions with unexpected windows network errors

### DISCUSSION THREAD

**Discussion by dougwood** [libvips/libvips#3138] [Q&A] [answered]
I am having intermittent problems saving bigTiff pyramid files using tiffsave() when the destination file is on an AWS Storage Gateway server backed volume.  This issue may also be happening when a NAS is used but I have only one report of that. I am using VIPS 8.13.2 on Windows (AWS EC2 instance or laptop). My python script is sometimes failing to save a particular image in a collection of similar images. All of the images are all being processed at once using the same pipeline running in separate threads.  

The place in the script where the error sometimes occurs is this:

    out.tiffsave(outfilepath, compression='lzw', xres=resolution, yres=resolution, resunit='cm', \
                 bigtiff=True, pyramid=True, tile=True, tile_width=256, tile_height=256)

and the returned messages usually start with:
CreateFileMapping failed: TIFFAppendToStrip: Maximum TIFF file size exceeded
vips2tiff: TIFF write tile failed
vips_mapfile: unable to CreateFileMapping
windows error: An unexpected network error occurred.

vips2tiff: TIFF write tile failed
vips_mapfile: unable to CreateFileMapping
windows error: An unexpected network error occurred.

I think the Maximum TIFF file size exceeded message may be erroneous as I am using bigtiff=True and bigtiff images are always being produced under normal circumstances.

Some observations:
- If I run exactly the same code on a laptop with locally mounted hard disks processing exactly the same images, I never see these errors.
- After a failure, if I try running the same processing again it will usually be successful, even when saving to the same Storage Gateway.
- In Windows Explorer, when one navigates to a new directory on the Storage Gateway volume, there is often a long delay while the Gateway server restores the files in the directory from S3 to EBS volumes and retrieves the data need to display the directory contents in the Windows Explorer window. We are using a Storage Gateway, BTW, for cost savings (S3 vs EBS) and to allow for volumes with essentially 'infinite' storage.

Now I realize that if a network error occurs or the file system simply won't cooperate, this is not VIPS' fault, but given that retrying the same processing again is usually successful, I wonder if VIPS could be modified to be more tolerant to these issues (increase timeouts or try the same operation again after some delay allowing the file system to catch-up?).

Failing a fix in VIPS tiffsave() for this issue, what would you recommend I try: catching these exceptions, then do something to restore the Storage Gateway cache or simply wait for it to catch up and try again?  Does this have something to do with my running several (5) threads at once saving different images (too much to ask of the file system?)

Thank you for your help and advice!
Doug

**Comment by jcupitt** [comment, marked as answer]
Hi @dougwood,

I think I would catch errors from pyramid write and retry with a timeout. Pyramid building is slow and needs a lot of resources, so all kinds of things can go wrong, unfortunately (as you say).

An alternative might be to make the pyramids to local storage, and then have a separate process for background uploads. Huge network writes are easily the most difficult, slow and unreliable part of the process, so that sounds like as win.

libvips 8.13 now (finally!) supports streamed write of TIFF pyramids, so you could also try that. It would push memory use up quite a bit (it has to buffer the low res levels in memory), but would save on local disc storage, and perhaps put less stress on the network connection.

**Comment by dougwood** [comment]
Thanks, as always, John.  I will try catching exceptions from each pyramid write and retry with a delay, just to see if that does help.  I am writing several of these pyramids at once (multichannel images), and that may be putting excessive load on the file system cache.  As I said, running the same processing again at a different time will sometimes work without a hitch.  And, yes, we were contemplating writing to a local store and then copying everything to the desired volume at the end, but that was a good bit of plumbing and infrastructure to set up and there still could be problems with the copy operation in the end.  I'll look into the streamed write of TIFF pyramids, too.  Cheers!
